src/compiler/checker.ts

```diff
@@ -20732,7 +20738,59 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
 
         for (let i = 0; i < paramCount; i++) {
             const sourceType = i === restIndex ? getRestOrAnyTypeAtPosition(source, i) : tryGetTypeAtPosition(source, i);
-            const targetType = i === restIndex ? getRestOrAnyTypeAtPosition(target, i) : tryGetTypeAtPosition(target, i);
+            let targetType = i === restIndex ? getRestOrAnyTypeAtPosition(target, i) : tryGetTypeAtPosition(target, i);
+            if (i === restIndex && targetType && sourceType && isTupleType(sourceType)) {
+                targetType = mapType(targetType, t => {
+                    if (
+                        !isTupleType(t) ||
+                        // When both sides are tuples of the same structure, we don't want to "propagate" types from elements of variable positions
+                        // to the following positions as that would disallow signatures of the exact same structures when trailing fixed elements are involved:
+                        //
+                        // let fn: (...rest: [...string[], number]) => void = (...rest: [...string[], number]) => {}; // ok
+                        //
+                        // Since we want to allow contextual types to flow into paremeters, we don't need to differentiate between rest and variadic elements
+                        // as that doesn't affect the contextual type of the parameter
+                        isTupleTypeStructureMatching(sourceType, t, TupleStructureComparisonKind.MatchVariable)
+                    ) {
+                        return t;
+                    }
+
+                    // We create a tuple type based on the target elements and the source's length here.
+                    // When the source signature accepts less parameters than the target signature
+                    // we only need to check the *used* elements of the target tuple, the rest is ignored by the source anyway
+                    // and thus it can be safely ignored here.
+                    //
+                    // let fn: (a: number, b: string) => void = (a: number) => {}; // ok
+                    //
+                    // In addition to that we also want to "propagate" element types of variable positions
+                    // to all following positions, as that represents possible argument types.
+                    //
+                    // function fn(...[a, b]: [...number[], string]) {
+                    //   a; // number | string
+                    //   b; // number | string
+                    // }
+                    // fn('str'); // valid
+                    const elementTypes: Type[] = [];
+                    const elementFlags: ElementFlags[] = [];

```

Comment:
Typically we use `append` rather than pushing onto an empty list directly; this is typically more efficient due to how the runtime constructs arrays.